File: backend/src/main/java/com/back/global/security/CustomAuthentication.java
package com.back.global.security;

import com.back.domain.member.entity.Member;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.GrantedAuthority;

import java.util.Collection;

public class CustomAuthentication implements Authentication {
    private final Authentication authentication;
    private final Member member;

    public CustomAuthentication(Authentication authentication, Member member) {
        this.authentication = authentication;
        this.member = member;
    }

    public Member getMember() {
        return member;
    }

    @Override
    public Collection<? extends GrantedAuthority> getAuthorities() {
        return authentication.getAuthorities();
    }

    @Override
    public Object getCredentials() {
        return authentication.getCredentials();
    }

    @Override
    public Object getDetails() {
        return authentication.getDetails();
    }

    @Override
    public Object getPrincipal() {
        return authentication.getPrincipal();
    }

    @Override
    public boolean isAuthenticated() {
        return authentication.isAuthenticated();
    }

    @Override
    public void setAuthenticated(boolean isAuthenticated) throws IllegalArgumentException {
        authentication.setAuthenticated(isAuthenticated);
    }

    @Override
    public String getName() {
        return authentication.getName();
    }
}

File: backend/src/main/java/com/back/global/security/CustomAuthentication.kt
package com.back.global.security

import com.back.domain.member.entity.Member
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken
import org.springframework.security.core.Authentication
import org.springframework.security.core.GrantedAuthority
import org.springframework.security.core.authority.SimpleGrantedAuthority

class CustomAuthentication(
    private val authentication: Authentication,
    val member: Member
) : Authentication by authentication {
    //클래스 위임으로 간단하게

    override fun getAuthorities(): Collection<GrantedAuthority> {
        return listOf(SimpleGrantedAuthority("ROLE_" + member.role.name))
    }

    fun getMember(): Member {
        return member
    }

    companion object {
        fun from(member: Member): CustomAuthentication {
            val authentication: Authentication = UsernamePasswordAuthenticationToken(
                member.email,
                null,  //비번
                listOf(SimpleGrantedAuthority("ROLE_" + member.role.name))
            )
            return CustomAuthentication(authentication, member)
        }
    }
}